File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/low_level/low_cmd.py
#!/usr/bin/env python3

import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time
import math


def main():
    print("=== Low Level Command Test ===")

    # Initialize objects
    status = h10w.H10wStatus()
    system = h10w.H10wSystem()
    low_level = h10w.H10wLowLevel()

    ret, move_msg = status.getMoveMessage()
    if ret != 0 or move_msg is None:
        print("获取运动信息失败，任务终止")
        exit(-1)

    left_arm_near_zero = all(
        abs(value - 0.0) < 0.001 for value in move_msg.position[:7]
    )
    if not left_arm_near_zero:
        print("为保证安全，请先将左臂回零")
        exit(-1)

    ret = system.powerControl(humanoid_sdk_py.PowerState.POWER_ON)
    if ret != 0:
        print("上电失败")
        exit(-1)

    ret = system.clearError()
    if ret != 0:
        print("清除错误失败")
        exit(-1)

    ret = system.upperEnableControl(humanoid_sdk_py.EnableState.ENABLE_ON)
    if ret != 0:
        print("上半身使能失败")
        exit(-1)

    ret = system.enableController(False)
    if ret != 0:
        print("禁用控制器失败")
        return -1

    # Set low level command for left arm joint 1
    MAX_POS = math.pi * 10.0 / 180.0
    DURATION = 10  # ms
    TOTAL_TIME = 2000  # ms
    control_count = TOTAL_TIME // DURATION
    low_cmd = h10w.LowCmd()

    try:
        for i in range(control_count):
            start_time = time.time()

            pos = math.sin(math.pi * i * DURATION / 1000.0) * MAX_POS
            vel = math.cos(math.pi * i * DURATION / 1000.0) * MAX_POS * math.pi
            joint_index = h10w.L_ARM_JOINT1 - 1
            low_cmd.motor_cmd[joint_index].mode = 1
            low_cmd.motor_cmd[joint_index].q = pos
            low_cmd.motor_cmd[joint_index].dq = vel

            if low_level.writeLowCmd(low_cmd) != 0:
                print("pub low cmd failed!")
                return 0
            elapsed_time = (time.time() - start_time) * 1000
            sleep_time = DURATION - elapsed_time
            if sleep_time > 0:
                time.sleep(sleep_time / 1000.0)
    finally:
        # Restart controller after publishing low commands
        system.enableController(True)

    print("Low level command test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/low_level/low_cmd.cpp
// low_cmd.cpp

#include <iostream>
#include <algorithm>
#include <thread>
#include <cmath>
#include <chrono>

#include "common/common_def.h"
#include "h10-w/h10w_low_level.h"
#include "h10-w/h10w_status.h"
#include "h10-w/h10w_system.h"

using namespace humanoid;
using namespace humanoid::h10w;

constexpr double MAX_POS = M_PI * 10.0 / 180.0;
constexpr double DURATION = 10;      // unit: ms
constexpr double TOTAL_TIME = 2000;  // unit: ms

int main() {
  std::cout << "=== Low Level Command Test ===" << std::endl;
  H10wStatus status;
  H10wSystem system;
  H10wLowLevel low_level;

  MoveMessageParams moveMsg;
  int result = status.getMoveMessage(moveMsg);
  if (result != 0) {
    std::cerr << "获取运动信息失败，任务终止" << std::endl;
    return -1;
  }
  bool leftArmNearZero =
      std::all_of(moveMsg.position.begin(), moveMsg.position.begin() + 7,
                  [](double value) { return value - 0.0 < 0.001; });
  if (!leftArmNearZero) {
    std::cerr << "为保证安全，请先将左臂回零" << std::endl;
    return -1;
  }
  result = system.powerControl(PowerState::POWER_ON);
  if (result != 0) {
    std::cerr << "上电失败" << std::endl;
    return -1;
  }
  result = system.clearError();
  if (result != 0) {
    std::cerr << "清错失败" << std::endl;
    return -1;
  }
  result = system.upperEnableControl(humanoid::EnableState::ENABLE_ON);
  if (result != 0) {
    std::cerr << "上半身使能失败" << std::endl;
    return -1;
  }
  result = system.enableController(false);
  if (result != 0) {
    std::cerr << "禁用控制器失败" << std::endl;
    return -1;
  }

  LowCmd low_cmd;
  // start control loop
  constexpr size_t control_count = TOTAL_TIME / DURATION;
  for (size_t i = 0; i < control_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();
    auto pos = std::sin(M_PI * i * DURATION / 1000.0) * MAX_POS;
    auto vel = std::cos(M_PI * i * DURATION / 1000.0) * MAX_POS * M_PI;
    low_cmd.joint_motor_cmd[static_cast<int>(JointIndexValue::L_ARM_JOINT1) - 1]
        .mode = 1;
    low_cmd.joint_motor_cmd[static_cast<int>(JointIndexValue::L_ARM_JOINT1) - 1]
        .q = pos;
    low_cmd.joint_motor_cmd[static_cast<int>(JointIndexValue::L_ARM_JOINT1) - 1]
        .dq = vel;
    if (low_level.writeLowCmd(low_cmd) != 0) {
      std::cout << "pub low cmd failed!" << std::endl;
      system.enableController(true);
      return 0;
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = end - start;
    std::this_thread::sleep_for(
        std::chrono::microseconds(static_cast<int>(DURATION * 1000)) - elapsed);
  }
  // restart controller after pub low cmd
  system.enableController(true);
  return 0;
}
